server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,362 @@
+     * @return true if shard is started on the search node, false otherwise
+     */
+    boolean isShardInWarmTier(final ShardRouting shard, final ClusterState clusterState) {
+        if (shard.unassigned()) {
+            return false;
+        }
+        final boolean isShardFoundOnSearchNode = clusterState.getNodes().get(shard.currentNodeId()).isSearchNode();
+        return shard.started() && isShardFoundOnSearchNode;
+    }
+
+    /**
+     * Completes the request level tiering for requestContext.
+     * @param requestContext tiering request context
+     */
+    void completeRequestLevelTiering(TieringRequestContext requestContext) {
+        tieringRequestContexts.remove(requestContext);
+        if (requestContext.getListener() != null) {
+            requestContext.getListener().onResponse(constructToHotToWarmTieringResponse(requestContext.getFailedIndices()));
+        }
+    }
+
+    TieringRequestContext findRequestContextForIndex(final Index index) {
+        return tieringRequestContexts.stream()
+            .filter(tieringRequestContext -> tieringRequestContext.hasIndex(index))
+            .findFirst()
+            .orElse(null);
+    }
+
+    /**
+     * Updates the request context for tiered indices,
+     * Moves tiered indices to successful state,
+     * Checks and completes the request level tiering
+     * @param indexTieringInfos set of tiered indices
+     */
+    void updateRequestContextForTieredIndices(final Set<TieringRequestContext.IndexTieringInfo> indexTieringInfos) {
+        for (TieringRequestContext.IndexTieringInfo indexTieringInfo : indexTieringInfos) {
+            Index tieredIndex = indexTieringInfo.getIndex();
+            TieringRequestContext tieringRequestContext = findRequestContextForIndex(tieredIndex);
+            assert tieringRequestContext != null;
+            assert tieringRequestContext.getIndexTieringInfo(tieredIndex) == indexTieringInfo;
+            indexTieringInfo.markSuccessful();
+            if (tieringRequestContext.isRequestProcessingComplete()) {
+                logger.info("[HotToWarmTiering] Tiering is completed for the request [{}]", tieringRequestContext);
+                completeRequestLevelTiering(tieringRequestContext);
+            }
+        }
+    }
+
+    /**
+     * Updates the index metadata with the tiering settings/metadata for an accepted index.
+     * Accepted index is an index to be tiered from hot to warm.
+     * @param metadataBuilder metadata builder
+     * @param routingTableBuilder routing builder
+     * @param indexTieringInfo index tiering info
+     */
+    void updateIndexMetadataForAcceptedIndex(
+        final Metadata.Builder metadataBuilder,
+        final RoutingTable.Builder routingTableBuilder,
+        final TieringRequestContext.IndexTieringInfo indexTieringInfo
+    ) {
+        final Index index = indexTieringInfo.getIndex();
+        final IndexMetadata indexMetadata = metadataBuilder.build().index(index);
```

Comment:
Instead of re-building the metadata here, we can pass in the previous metadata object